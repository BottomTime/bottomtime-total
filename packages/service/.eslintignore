coverage/
dist/

src/types/**/*.d.ts
