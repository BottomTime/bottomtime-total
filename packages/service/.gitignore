/assets/templates/
/docs/swagger.json
