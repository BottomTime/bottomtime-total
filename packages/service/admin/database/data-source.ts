import path from 'path';
import { DataSource } from 'typeorm';

export async function getDataSource(
  url: string,
  requireSsl: boolean,
): Promise<DataSource> {
  const client = new DataSource({
    type: 'postgres',
    url,
    entities: [path.resolve(__dirname, '../../src/data/**/*.entity.ts')],
    ssl: requireSsl,
  });
  return await client.initialize();
}
