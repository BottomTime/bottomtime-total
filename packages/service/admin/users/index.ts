import { CommandModule } from 'yargs';

import { createAdmin } from './create-admin';
import { getUserToken } from './get-user-token';

export const userModule: CommandModule<{
  'postgres-uri': string;
  'require-ssl': boolean;
}> = {
  command: 'user',

  describe: 'Commands for managing a user account',

  builder(yargs) {
    return yargs
      .command(
        'new-admin',
        'Creates a new admin user.',
        (yargs) => {
          return yargs
            .positional('username', {
              alias: 'u',
              description: "The user's username.",
              type: 'string',
            })
            .positional('password', {
              alias: 'p',
              description: 'The password to set on the new admin account',
              type: 'string',
            })
            .help();
        },
        async (yargs) => {
          await createAdmin({
            postgresRequireSsl: yargs.requireSsl,
            potgresUri: yargs.postgresUri,
            username: yargs.username,
            password: yargs.password,
          });
        },
      )
      .command(
        'token <user>',
        'Gets a token for a user',
        (yargs) => {
          return yargs
            .positional('user', {
              demandOption: true,
              description: "The user's username or email address.",
              type: 'string',
            })
            .help();
        },
        async (yargs) => {
          await getUserToken(yargs.postgresUri, yargs.requireSsl, yargs.user);
        },
      )
      .help();
  },

  async handler() {
    /* TODO */
  },
};
