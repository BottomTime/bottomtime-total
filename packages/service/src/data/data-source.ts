import path from 'path';
import { DataSource, DataSourceOptions } from 'typeorm';

import { Config } from '../config';

export const PostgresDataSourceOptions: DataSourceOptions = {
  type: 'postgres',
  url: Config.postgresUri,
  entities: [path.join(__dirname, './**/*.entity.ts')],
  synchronize: false,
};

export async function initDataSource(): Promise<DataSource> {
  const AppDataSource = new DataSource(PostgresDataSourceOptions);
  return await AppDataSource.initialize();
}
