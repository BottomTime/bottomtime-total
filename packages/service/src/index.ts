import { Config } from './config';
import { createApp } from './create-app';
import { createLogger } from './logger';

const log = createLogger(Config.logLevel);

createApp(log)
  .then(async (app) => {
    await app.listen(Config.port);

    log.info(
      `ðŸŽ‰ Service has successfully started and is listening on port ${Config.port}. ðŸŽ‰`,
    );
  })
  .catch((error) => {
    log.fatal('[SERVICE] Failed to start service... Shutting down...', error);
    process.exit(1);
  });
