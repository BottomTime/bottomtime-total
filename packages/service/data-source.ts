import path from 'path';
import { DataSource } from 'typeorm';

import { PostgresDataSourceOptions } from './src/data/data-source';

export default new DataSource({
  ...PostgresDataSourceOptions,
  migrations: [path.join(__dirname, './migrations/*.ts')],
});
