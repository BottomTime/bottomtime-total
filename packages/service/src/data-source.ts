import 'reflect-metadata';
import { DataSourceOptions } from 'typeorm';

import { Config } from './config';
import * as Entities from './data';

export const PostgresDataSourceOptions: DataSourceOptions = {
  type: 'postgres',
  url: Config.postgresUri,
  entities: Object.values(Entities),
  synchronize: false,
};
