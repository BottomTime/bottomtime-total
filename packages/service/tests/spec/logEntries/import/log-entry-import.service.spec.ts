import { writeFile } from 'fs/promises';
import { resolve } from 'path';
import { IsNull, LessThanOrEqual, Repository } from 'typeorm';
import * as uuid from 'uuid';

import {
  LogEntryEntity,
  LogEntryImportEntity,
  UserEntity,
} from '../../../../src/data';
import { LogEntryImportFactory } from '../../../../src/logEntries/import/log-entry-import-factory';
import { LogEntryImportService } from '../../../../src/logEntries/import/log-entry-import.service';
import { User, UserFactory } from '../../../../src/users';
import { dataSource } from '../../../data-source';
import LogEntryData from '../../../fixtures/log-entries.json';
import TestData from '../../../fixtures/log-entry-imports.json';
import {
  createTestLogEntryImport,
  parseLogEntryImportJSON,
  parseLogEntryJSON,
} from '../../../utils';
import { createTestUser } from '../../../utils/create-test-user';

jest.mock('uuid');

const Now = new Date('2024-11-12T10:07:21-05:00');
const OwnerData: Partial<UserEntity> = {
  id: '114c1546-216b-4b58-873b-0c63a80fe04e',
  username: 'testuser',
};
const OtherUserData: Partial<UserEntity> = {
  id: '4c5be9b4-782d-45f8-bae1-19dfb120905e',
  username: 'otheruser',
};

describe('Log Entry Import Service', () => {
  let Entries: Repository<LogEntryEntity>;
  let Imports: Repository<LogEntryImportEntity>;
  let Users: Repository<UserEntity>;
  let service: LogEntryImportService;
  let importFactory: LogEntryImportFactory;

  let ownerData: UserEntity;
  let otherUserData: UserEntity;
  let owner: User;
  let testData: LogEntryImportEntity[];

  beforeAll(() => {
    Entries = dataSource.getRepository(LogEntryEntity);
    Imports = dataSource.getRepository(LogEntryImportEntity);
    Users = dataSource.getRepository(UserEntity);

    const userFactory = new UserFactory(Users);
    importFactory = new LogEntryImportFactory(Imports, Entries, userFactory);
    service = new LogEntryImportService(Imports, Entries, importFactory);
  });

  beforeEach(async () => {
    jest.useFakeTimers({
      now: Now,
      doNotFake: ['nextTick', 'setImmediate'],
    });

    ownerData = createTestUser(OwnerData);
    owner = new User(Users, ownerData);
    otherUserData = createTestUser(OtherUserData);
    await Users.save([ownerData, otherUserData]);

    testData = TestData.map((data) =>
      parseLogEntryImportJSON(
        data,
        data.owner.username === ownerData.username ? ownerData : otherUserData,
      ),
    );
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it.skip('generate test data', async () => {
    const imports = new Array<LogEntryImportEntity>(40);

    for (let i = 0; i < imports.length; i++) {
      const userEntity = i < 35 ? ownerData : otherUserData;
      imports[i] = {
        ...createTestLogEntryImport(i < 35 ? ownerData : otherUserData),
        owner: {
          id: userEntity.id,
          username: userEntity.username,
        } as UserEntity,
      };
    }

    await writeFile(
      resolve(__dirname, '../../fixtures/log-entry-imports.json'),
      JSON.stringify(imports, null, 2),
      'utf-8',
    );
  });

  describe('when initializing a new import', () => {
    it('will initialize a new import with minimal params and return it', async () => {
      const id = '5f207df6-5370-4dc5-baa2-2eb60bbfd9d0';
      jest.spyOn(uuid, 'v7').mockReturnValue(id);
      const result = await service.createImport(owner);
      expect(result.toJSON()).toEqual({
        id,
        owner: owner.username,
        date: Now,
        finalized: false,
      });

      const saved = await Imports.findOneOrFail({
        where: { id },
        relations: ['owner'],
      });
      expect(saved.owner.id).toEqual(owner.id);
      expect(saved.finalized).toBeNull();
      expect(saved.date).toEqual(Now);
      expect(saved.device).toBeNull();
      expect(saved.deviceId).toBeNull();
      expect(saved.bookmark).toBeNull();
    });

    it('will initialize a new import with all params and return it', async () => {
      const id = '27bcdbae-a2bb-4cee-a15e-9199a8874f81';
      const device = 'Suunto Kajigger';
      const deviceId = '1234567890';
      const bookmark = 'bookmark_xyz123';

      jest.spyOn(uuid, 'v7').mockReturnValue(id);
      const result = await service.createImport(owner, {
        device,
        deviceId,
        bookmark,
      });
      expect(result.toJSON()).toEqual({
        id,
        owner: owner.username,
        date: Now,
        finalized: false,
        device,
        deviceId,
        bookmark,
      });

      const saved = await Imports.findOneOrFail({
        where: { id },
        relations: ['owner'],
      });
      expect(saved.owner.id).toEqual(owner.id);
      expect(saved.finalized).toBeNull();
      expect(saved.date).toEqual(Now);
      expect(saved.device).toEqual(device);
      expect(saved.deviceId).toEqual(deviceId);
      expect(saved.bookmark).toEqual(bookmark);
    });
  });

  describe('when retrieving a single import', () => {
    it('will return the requested import', async () => {
      const data = createTestLogEntryImport(ownerData);
      const expected = importFactory.createImport(data);
      await Imports.save(data);

      const actual = await service.getImport(data.id);
      expect(actual?.toJSON()).toEqual(expected.toJSON());
    });

    it('will return undefined if the indicated import does not exist', async () => {
      await expect(
        service.getImport('ce41d20d-8eaa-4eb6-b60f-2afdf1ac8779'),
      ).resolves.toBeUndefined();
    });
  });

  describe('when listing imports', () => {
    let logEntryData: LogEntryEntity[];

    beforeEach(async () => {
      logEntryData = LogEntryData.map((data) =>
        parseLogEntryJSON(data, otherUserData),
      );
      for (let i = 0; i < testData.length; i++) {
        for (let j = 0; j < 5; j++) {
          logEntryData[i * 5 + j].owner = testData[i].owner;
          logEntryData[i * 5 + j].import = testData[i];
        }
      }

      await Users.save([ownerData, otherUserData]);
      await Imports.save(testData);
      await Entries.save(logEntryData);
    });

    it('will return an empty set if no imports match the criteria', async () => {
      const someOtherUser = new User(Users, createTestUser());
      const results = await service.listImports({
        owner: someOtherUser,
      });

      expect(results).toEqual({
        data: [],
        totalCount: 0,
      });
    });

    it('will return result set with finalized imports filtered out', async () => {
      const expectedLength = testData.filter(
        (i) => i.owner.id === OwnerData.id && i.finalized === null,
      ).length;

      const results = await service.listImports({
        owner,
        showFinalized: false,
      });

      expect(results.totalCount).toBe(expectedLength);
      expect(
        results.data.map((i) => ({
          owner: i.owner.username,
          date: i.date,
          finalized: i.finalized,
        })),
      ).toMatchSnapshot();
    });

    it('will not return results belonging to a different owner', async () => {
      const results = await service.listImports({
        owner,
        showFinalized: true,
        skip: 0,
        limit: 500,
      });
      expect(results.data.every((i) => i.owner.id === ownerData.id)).toBe(true);
    });

    it('will return result set with all results matching owner', async () => {
      const expectedLength = testData.filter(
        (i) => i.owner.id === ownerData.id,
      ).length;
      const results = await service.listImports({
        owner,
        showFinalized: true,
      });
      expect(results.data).toHaveLength(expectedLength);
      expect(results.totalCount).toBe(expectedLength);
      expect(
        results.data.map((i) => ({
          owner: i.owner.username,
          date: i.date,
          finalized: i.finalized,
        })),
      ).toMatchSnapshot();
    });

    it('will return paged results', async () => {
      const results = await service.listImports({
        owner,
        showFinalized: true,
        skip: 5,
        limit: 5,
      });

      expect(results.data).toHaveLength(5);
      expect(results.totalCount).toBe(35);
      expect(
        results.data.map((i) => ({
          owner: i.owner.username,
          date: i.date,
          finalized: i.finalized,
        })),
      ).toMatchSnapshot();
    });
  });

  describe('when expiring imports', () => {
    beforeEach(async () => {
      await Imports.save(testData);
    });

    it('will expire all non-finalized imports older than the indicated date', async () => {
      const expired = testData
        .filter((i) => i.finalized === null)
        .sort((a, b) => b.date.valueOf() - a.date.valueOf());
      const expiration = expired[expired.length - 3].date;
      const removed = await service.expireImports(expiration);
      expect(removed).toBe(expired.length - 4);

      await expect(Imports.count()).resolves.toBe(testData.length - removed);
      await expect(
        Imports.existsBy({
          finalized: IsNull(),
          date: LessThanOrEqual(expiration),
        }),
      ).resolves.toBe(false);
    });
  });
});
