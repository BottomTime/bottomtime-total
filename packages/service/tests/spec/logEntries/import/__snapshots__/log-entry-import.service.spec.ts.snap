// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Log Entry Import Service when listing imports will return paged results 1`] = `
[
  {
    "date": 2024-05-20T21:45:17.961Z,
    "finalized": false,
    "owner": "testuser",
  },
  {
    "date": 2024-11-12T05:20:31.505Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-10-24T19:26:33.797Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-10-11T08:12:02.577Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-10-09T13:10:30.383Z,
    "finalized": true,
    "owner": "testuser",
  },
]
`;

exports[`Log Entry Import Service when listing imports will return result set with all results matching owner 1`] = `
[
  {
    "date": 2024-10-30T00:44:50.840Z,
    "finalized": false,
    "owner": "testuser",
  },
  {
    "date": 2024-10-09T01:51:46.636Z,
    "finalized": false,
    "owner": "testuser",
  },
  {
    "date": 2024-08-15T17:20:30.544Z,
    "finalized": false,
    "owner": "testuser",
  },
  {
    "date": 2024-06-28T15:51:44.543Z,
    "finalized": false,
    "owner": "testuser",
  },
  {
    "date": 2024-05-22T06:21:46.799Z,
    "finalized": false,
    "owner": "testuser",
  },
  {
    "date": 2024-05-20T21:45:17.961Z,
    "finalized": false,
    "owner": "testuser",
  },
  {
    "date": 2024-11-12T05:20:31.505Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-10-24T19:26:33.797Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-10-11T08:12:02.577Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-10-09T13:10:30.383Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-10-08T11:49:58.856Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-10-06T22:16:17.486Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-10-01T22:21:49.989Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-10-01T06:56:04.708Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-09-20T23:46:14.207Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-09-20T08:10:03.645Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-09-03T18:03:26.149Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-09-02T12:09:25.472Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-09-01T08:48:52.581Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-08-29T00:04:55.257Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-08-21T21:20:29.984Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-08-16T16:13:23.402Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-08-10T14:15:56.594Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-07-29T05:12:50.929Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-07-19T15:06:47.597Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-07-19T03:28:39.842Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-07-18T00:11:44.426Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-07-11T19:57:33.174Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-07-07T21:45:42.949Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-06-24T03:43:03.900Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-06-23T10:42:38.083Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-06-21T17:49:07.151Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-06-06T03:41:42.098Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-06-01T03:53:59.529Z,
    "finalized": true,
    "owner": "testuser",
  },
  {
    "date": 2024-05-31T06:05:33.958Z,
    "finalized": true,
    "owner": "testuser",
  },
]
`;

exports[`Log Entry Import Service when listing imports will return result set with finalized imports filtered out 1`] = `
[
  {
    "date": 2024-10-30T00:44:50.840Z,
    "finalized": false,
    "owner": "testuser",
  },
  {
    "date": 2024-10-09T01:51:46.636Z,
    "finalized": false,
    "owner": "testuser",
  },
  {
    "date": 2024-08-15T17:20:30.544Z,
    "finalized": false,
    "owner": "testuser",
  },
  {
    "date": 2024-06-28T15:51:44.543Z,
    "finalized": false,
    "owner": "testuser",
  },
  {
    "date": 2024-05-22T06:21:46.799Z,
    "finalized": false,
    "owner": "testuser",
  },
  {
    "date": 2024-05-20T21:45:17.961Z,
    "finalized": false,
    "owner": "testuser",
  },
]
`;
